[interfaces] add pareto interfaces

alxlyj opened the pull request:
## Project Robyn

**Depends on https://github.com/facebookexperimental/Robyn/pull/1050**

**Summary:** We are including the interfaces (methods and data classes) for the pareto subcomponent in Python. This interface is inferred from pareto.R.  

## What is new:
- feat: New Python interfaces for pareto subcomponent. 

## Test Plan

### Class diagram
```mermaid
classDiagram
    class ParetoOptimizer {
        -MMMData mmm_data
        -ModelOutputs model_outputs
        -ResponseCurveCalculator response_calculator
        -ImmediateCarryoverCalculator carryover_calculator
        -ParetoUtils pareto_utils
        +optimize(pareto_fronts: str, min_candidates: int, calibration_constraint: float, calibrated: bool) ParetoResult
        -_aggregate_model_data(calibrated: bool) Dict
        -_compute_pareto_fronts(data: Dict, pareto_fronts: str, min_candidates: int) DataFrame
        -_compute_response_curves(pareto_fronts_df: DataFrame) Dict
        -_generate_plot_data(pareto_fronts_df: DataFrame, response_curves: Dict) Dict
    }
    class ParetoResult {
        +List[str] pareto_solutions
        +int pareto_fronts
        +DataFrame result_hyp_param
        +DataFrame x_decomp_agg
        +DataFrame result_calibration
        +DataFrame media_vec_collect
        +DataFrame x_decomp_vec_collect
        +Dict[str, DataFrame] plot_data_collect
        +DataFrame df_caov_pct_all
    }
    class ResponseCurveCalculator {
        -MMMData mmm_data
        -ModelOutputs model_outputs
        +calculate_response(model_id: str, metric_name: str, channel: str, date_range: Tuple[str, str]) ResponseCurveData
        +calculate_all_responses(model_id: str) List[ResponseCurveData]
        -_get_hill_params(model_id: str, channel: str) HillParameters
        -_get_spend_values(channel: str, date_range: Tuple[str, str]) ndarray
        -_calculate_response_values(hill_params: HillParameters, spend_values: ndarray) ndarray
        -_get_model_channels(model_id: str) List[str]
    }
    class ImmediateCarryoverCalculator {
        -MMMData mmm_data
        -ModelOutputs model_outputs
        +calculate(sol_id: Optional[str], start_date: Optional[str], end_date: Optional[str]) List[EffectDecomposition]
        +calculate_all() DataFrame
        -_get_date_range(start_date: Optional[str], end_date: Optional[str]) DateRange
        -_calculate_saturated_dataframes(sol_id: str) Dict[str, DataFrame]
        -_calculate_decomposition(sol_id: str, saturated_dfs: Dict[str, DataFrame], date_range: DateRange) Dict[str, Dict[str, ndarray]]
        -_aggregate_effects(decomp_data: Dict[str, Dict[str, ndarray]], date_range: DateRange) List[EffectDecomposition]
        -_get_default_solution_id() str
        -_get_all_solution_ids() List[str]
    }
    class ParetoUtils {
        +calculate_pareto_front(x: ndarray, y: ndarray, max_fronts: int) DataFrame
        +calculate_error_scores(result_hyp_param: DataFrame, ts_validation: bool) ndarray
        +calculate_nrmse(y_true: ndarray, y_pred: ndarray) float
        +calculate_mape(y_true: ndarray, y_pred: ndarray) float
        +calculate_decomp_rssd(decomp_values: ndarray) float
        +find_knee_point(x: ndarray, y: ndarray) Tuple[float, float]
        +calculate_hypervolume(points: ndarray) float
        +normalize_objectives(objectives: ndarray) ndarray
        +calculate_crowding_distance(points: ndarray) ndarray
    }
    class DateRange {
        +datetime start_date
        +datetime end_date
        +int start_index
        +int end_index
    }
    class EffectDecomposition {
        +str channel
        +float immediate_effect
        +float carryover_effect
        +float total_effect
    }
    class HillParameters {
        +float alpha
        +float gamma
    }
    class ResponseCurveData {
        +str model_id
        +str metric_name
        +str channel
        +ndarray spend_values
        +ndarray response_values
        +HillParameters hill_params
    }
    ParetoOptimizer --> ParetoResult : produces
    ParetoOptimizer --> ResponseCurveCalculator : uses
    ParetoOptimizer --> ImmediateCarryoverCalculator : uses
    ParetoOptimizer --> ParetoUtils : uses
    ParetoOptimizer ..> MMMData : uses
    ParetoOptimizer ..> ModelOutputs : uses
    ResponseCurveCalculator ..> MMMData : uses
    ResponseCurveCalculator ..> ModelOutputs : uses
    ResponseCurveCalculator --> ResponseCurveData : produces
    ResponseCurveCalculator --> HillParameters : uses
    ImmediateCarryoverCalculator ..> MMMData : uses
    ImmediateCarryoverCalculator ..> ModelOutputs : uses
    ImmediateCarryoverCalculator --> EffectDecomposition : produces
    ImmediateCarryoverCalculator --> DateRange : uses
```

### Parity:
```mermaid
graph TB
    subgraph R["R Implementation"]
        R1[robyn_pareto]
        R2[pareto_front]
        R3[robyn_immcarr]
    end
    subgraph Python["Python Implementation"]
        P1[ParetoOptimizer.optimize]
        P2[ParetoUtils.calculate_pareto_front]
        P3[ImmediateCarryoverCalculator.calculate]
    end
    %% Connections
    R1 ---- P1
    R2 ---- P2
    R3 ---- P3
    %% New classes and data structures
    P4[ParetoResult]
    P5[ResponseCurveCalculator]
    P6[DateRange]
    P7[EffectDecomposition]
    P8[HillParameters]
    P9[ResponseCurveData]
    P10[ParetoUtils]
    %% Connections to new classes
    P1 --> P4
    P1 --> P5
    P1 --> P10
    P3 --> P6
    P3 --> P7
    P5 --> P8
    P5 --> P9
    %% Legend
    classDef implemented fill:#90EE90,stroke:#006400, color:black, stroke-width:2px;
    classDef new fill:#ADD8E6,stroke:#00008B, color:black, stroke-width:2px;
    class R1,R2,R3,P1,P2,P3 implemented;
    class P4,P5,P6,P7,P8,P9,P10 new;
    legendA[Implemented]
    legendB[New Python Classes/Structures]
    class legendA implemented;
    class legendB new;
```

-  No implementations have been included just yet with this PR so it would be lightweight to land. More rigorous testing would be included in the implementation PR.

sumane81 commented on python/src/robyn/modeling/pareto/immediate_carryover_calculator.py:
```diff
@@ -0,0 +1,108 @@
+        output_collect (ModelOutputs): Output data from the model runs.
+    """
+
+    def __init__(self, input_collect: MMMData, output_collect: ModelOutputs):
```
please use meaningful and relevant variable names. For example, mmmdata and model_outputs.

sumane81 commented on python/src/robyn/modeling/pareto/immediate_carryover_calculator.py:
```diff
@@ -0,0 +1,108 @@
+        # Implementation here
+        pass
+
+    def _get_date_range(self, start_date: Optional[str], end_date: Optional[str]) -> Dict[str, Any]:
```
Recommend using datetime as input than string, which you will have to convert to datetime.

sumane81 commented on python/src/robyn/modeling/pareto/immediate_carryover_calculator.py:
```diff
@@ -0,0 +1,108 @@
+            end_date (Optional[str]): End date for the calculation.
+
+        Returns:
+            Dict[str, Any]: Dictionary containing start and end dates and indices.
```
Avoid using open ended dictionary. Either create complex object where it makes sense. Otherwise, use TypedDictionary

sumane81 replied:
Also, let's not use Any if possible.

sumane81 commented on python/src/robyn/modeling/pareto/pareto_optimizer.py:
```diff
@@ -0,0 +1,131 @@
+        carryover_calculator (ImmediateCarryoverCalculator): Calculator for immediate and carryover effects.
+    """
+
+    def __init__(self, input_collect: MMMData, output_models: ModelOutputs):
```
Use relevant parameter names.

sumane81 commented on python/src/robyn/modeling/pareto/pareto_optimizer.py:
```diff
@@ -0,0 +1,131 @@
+        pareto_fronts: str = "auto",
+        min_candidates: int = 100,
+        calibration_constraint: float = 0.1,
+        quiet: bool = False,
```
Let's remove quiet. It should be managed at logger level.

sumane81 commented on python/src/robyn/modeling/pareto/pareto_optimizer.py:
```diff
@@ -0,0 +1,131 @@
+        # Implementation here
+        pass
+
+    def _prepare_data(self) -> Dict[str, pd.DataFrame]:
```
It's unclear on what this method does. Can we be specific on what preparations it does?

sumane81 commented on python/src/robyn/modeling/pareto/pareto_utils.py:
This doesn't seem to be a class? if functions are self independent then you can use class and define these methods as static methods.

sumane81 commented on python/src/robyn/modeling/pareto/__init__.py:
Do we need this file here?

sumane81 commented on python/src/robyn/modeling/pareto/immediate_carryover_calculator.py:
```diff
@@ -0,0 +1,108 @@
+# pyre-strict
+
+from typing import Optional, Dict, Any
```
Let's not use Any. If you absolutely need to, please justify.

sumane81 commented on python/src/robyn/modeling/pareto/immediate_carryover_calculator.py:
```diff
@@ -0,0 +1,108 @@
+            input_collect (MMMData): Input data for the marketing mix model.
+            output_collect (ModelOutputs): Output data from the model runs.
+        """
+        self.input_collect = input_collect
```
Same as above, no input_collect, output_collect.

sumane81 commented on python/src/robyn/modeling/pareto/pareto_optimizer.py:
```diff
@@ -0,0 +1,131 @@
+# pyre-strict
+
+from typing import List, Dict, Any, Optional
```
Let's not use Any, if possible.

sumane81 commented on python/src/robyn/modeling/pareto/pareto_optimizer.py:
```diff
@@ -0,0 +1,131 @@
+from robyn.modeling.entities.modeloutputs import ModelOutputs
+from robyn.modeling.pareto.response_curve_calculator import ResponseCurveCalculator
+from robyn.modeling.pareto.immediate_carryover_calculator import ImmediateCarryoverCalculator
+from robyn.modeling.pareto.pareto_utils import pareto_front
```
pareto_utils should be class

sumane81 commented on python/src/robyn/modeling/pareto/pareto_optimizer.py:
```diff
@@ -0,0 +1,131 @@
+    Pareto front calculation, response curve calculation, and plot data preparation.
+
+    Attributes:
+        input_collect (MMMData): Input data for the marketing mix model.
```
use mmm_data or mmmdata as variable name. Let's be consistent with other parts of the codebase

sumane81 commented on python/src/robyn/modeling/pareto/pareto_optimizer.py:
```diff
@@ -0,0 +1,131 @@
+
+    Attributes:
+        input_collect (MMMData): Input data for the marketing mix model.
+        output_models (ModelOutputs): Output data from the model runs.
```
model_outputs instead of output_models, just to be consistent

sumane81 reviewed (changes requested):
Mostly nits; but strongly suggest addressing them.

alxlyj commented:
### pareto_optimizer.py
1. We've removed the use of Any type where possible, using more specific types instead.
1. pareto_utils is now a class (ParetoUtils) rather than a module with standalone functions.
1. We've used mmm_data consistently as the variable name for the input data.
1. We've changed output_models to model_outputs for consistency.
1. We've renamed input_collect to mmm_data for clarity.
1. We've removed the quiet parameter, assuming logging will be handled at a different level.
1. We've renamed _prepare_data to _aggregate_model_data and provided a more detailed description of what it does.
1. We've added comprehensive docstrings for the class and all methods, explaining their purpose, parameters, and return values in detail.

alxlyj commented:
Updated diagrams.
